### DISCUSSION TITLE
Oracle dblink usage with ORM (Table object)

### DISCUSSION THREAD

**Discussion by SincerelyUnique** [sqlalchemy/sqlalchemy#7717] [Usage Questions]
### Describe the use case

Good afternoon, sorry to bother you here, I have a question about dblink usage, could you help me check? thanks you~

I want to query oracle db with dblink but I don't know how to create `Table` object, I also tried many times but failed, I know how to query without dblink, the code like below and works:

```
from sqlalchemy import func, Table, MetaData, Column, String, Integer, DateTime, select, create_engine
from sqlalchemy.pool import NullPool
from sqlalchemy.sql.elements import and_
import pandas as pd

engine = create_engine('xxx', poolclass=NullPool, echo='debug')


xxxx_xxxx_test = Table(
    'xxxx_xxxx_test',
    MetaData(),
    Column('code', Integer, primary_key=True),
    Column('text', String(20)),
    Column('createtime', DateTime, default=func.now()),
    Column('text2', String(20))
)


if __name__ == '__main__':
    stmt1 = xxxx_xxxx_test.select().where(xxxx_xxxx_test.columns.code == 2)
    data1 = engine.execute(stmt1).fetchall()
    df1 = pd.read_sql_query(stmt1, engine)

    stmt2 = xxxx_xxxx_test.select().\
        where(and_(xxxx_xxxx_test.columns.code == 1,
                   xxxx_xxxx_test.columns.text == 'jalen-test-query2'))
    data2 = engine.execute(stmt2).fetchall()
    df2 = pd.read_sql_query(stmt2, engine)

    stmt3 = select([xxxx_xxxx_test.columns.code,
                    xxxx_xxxx_test.columns.text,
                    xxxx_xxxx_test.columns.createtime]).\
        where(and_(xxxx_xxxx_test.columns.code == 1,
                   xxxx_xxxx_test.columns.text == 'jalen-test-query2'))
    data3 = engine.execute(stmt3).fetchall()
    df3 = pd.read_sql_query(stmt3, engine)

    stmt4 = select([xxxx_xxxx_test.columns.text])
    data4 = engine.execute(stmt4).fetchall()
    df4 = pd.read_sql_query(stmt4, engine)
    print(1)
```

but when I change the tablename to dblink, it not works, I found document from https://docs.sqlalchemy.org/en/14/dialects/oracle.html#synonym-dblink-reflection and try like this:

```
option = Table(
    'some_table@some_link',
    MetaData(),
    Column('id', String(32)),
    Column('optionname', String(128)),
    Column('description', String(128)),
    Column('createtime', DateTime),
    Column('lastmodifiedtime', DateTime),
    Column('type', String(32)),
    autoload_with=engine,
    oracle_resolve_synonyms=True)


if __name__ == '__main__':
    stmt1 = option.select().where(option.columns.id == 'xxx')
    data1 = engine.execute(stmt1).fetchall()
    df1 = pd.read_sql_query(stmt1, engine)
    print(1)
```

It not works and alert exception message "sqlalchemy.exc.NoSuchTableError"

### Databases / Backends / Drivers targeted

oracle 19c

### Example Use

```
option = Table(
    'some_table@some_link',
    MetaData(),
    Column('id', String(32)),
    Column('optionname', String(128)),
    Column('description', String(128)),
    Column('createtime', DateTime),
    Column('lastmodifiedtime', DateTime),
    Column('type', String(32)),
    autoload_with=engine,
    oracle_resolve_synonyms=True)


if __name__ == '__main__':
    stmt1 = option.select().where(option.columns.id == 'xxx')
    data1 = engine.execute(stmt1).fetchall()
    df1 = pd.read_sql_query(stmt1, engine)
    print(1)
```

### Additional context

The db_link not in all_synonyms but in dba_db_links
sqlalchemy version : 1.4.29

My English not very good! :)

**Comment by zzzeek** [comment]
we dont get many requests for DBLINK support so the API is very limited.  

the two options for using DBLINK are:

1. **create synonyms in the Oracle database** - here, if you wanted to query from ``some_table@some_link``, you would create a synonym first:

```sql
sqlplus> CREATE SYNONYM some_table FOR some_table@some_link
```

Then you would refer to your table as ``Table("some_table", metadata, ...)``


2. **Use a special "unquoted" name (not tested)**

The DBLINK name can be formatted with the table as two separate tokens that won't be quoted by using the [quoted_name](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=quoted_name#sqlalchemy.sql.expression.quoted_name) construct.  Oracle probably doesn't like the "some_table@some_link" being rendered over and over again, so to reduce the rendering of that full name you can then use an alias, and that's what you'd use for statements.  The below example illustrates this:

```py
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.dialects import oracle
from sqlalchemy.sql.expression import quoted_name

option = Table(
    quoted_name("some_table@some_link", quote=False),
    MetaData(),
    Column("id", String(32)),
    Column("optionname", String(128)),
    Column("description", String(128)),
    Column("createtime", DateTime),
    Column("lastmodifiedtime", DateTime),
    Column("type", String(32)),
)


option_alias = option.alias("some_table")

stmt = option_alias.select().where(option_alias.columns.id == "xxx")

# to execute the statement
result = some_connection.execute(stmt)

# or to print it locally
print(stmt.compile(dialect=oracle.dialect()))

```

This will render a statement that Oracle should be OK with:

```sql
SELECT some_table.id, some_table.optionname, some_table.description, some_table.createtime, some_table.lastmodifiedtime, some_table.type 
FROM some_table@some_link some_table 
WHERE some_table.id = :id_1
```

Note also that the table above has the `Column` objects present; you don't need to use "autoload" here.


3. We likely need to add a "dblink" option anyway for other work we are doing in 2.0 cc @CaselIT , I'll create a task for that.

**Reply by CaselIT** [reply]
Yes at the moment there is a private execution option in the wip multiple reflection. But we can create a public api for it I guess

**Reply by zzzeek** [reply]
I created #7718 earlier today.   it's more of a quick fix.

**Reply by SincerelyUnique** [reply]
Thank you so much! @zzzeek  @CaselIT  I will have a try with your second option~~
